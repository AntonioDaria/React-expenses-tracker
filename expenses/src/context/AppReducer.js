//  4 We need a Reducer to specify the application state changes to our store(context) in response to certain actions

export const appContextReducer = (state, action) => {

    switch (action.type) {
        case 'DELETE_TRANSACTION':
            return {
                ...state,
                transactions: state.transactions.filter(transaction => transaction.id !== action.payload)
            }
        case 'ADD_TRANSACTION':
            return {
                ...state,
                transactions: [action.payload, ...state.transactions]
            }
        default:
            return state;
    };
}

export default appContextReducer