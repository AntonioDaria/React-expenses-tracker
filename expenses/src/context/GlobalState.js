import React, { createContext, useReducer } from 'react'
import appContextReducer from './AppReducer'

// 1 Initial State

export const INITIAL_TRANSACTION_CONTEXT_STATE = {
    transactions: [ 
       
    ]
}

// 2 Create Context

export const GlobalContext = createContext(INITIAL_TRANSACTION_CONTEXT_STATE);



//  3 In order for our components to have access to our global state we need a Provider

//Provider Component

export const GlobalProvider = ({ children }) => {

    const [state, dispatch] = useReducer(appContextReducer, INITIAL_TRANSACTION_CONTEXT_STATE);

//Actions to make calls to reducer

    function deleteTransaction(id) {
        dispatch({
            type: 'DELETE_TRANSACTION',
            payload: id
        });
    }

    function addTransaction(transaction) {
        dispatch({
            type: 'ADD_TRANSACTION',
            payload: transaction
        });
    }

    return (
        <GlobalContext.Provider
            value={{ transactions: state.transactions,
            deleteTransaction,
            addTransaction
            }}>
            {children}
        </GlobalContext.Provider>
    )

}

