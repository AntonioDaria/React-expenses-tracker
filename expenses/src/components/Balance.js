import React, { useContext } from 'react'
import { GlobalContext } from '../context/GlobalState'

export const Balance = () => {
  const { transactions } = useContext(GlobalContext)


  const arrayBalance = transactions.map(transaction => { return transaction.amount })
  const total = arrayBalance.reduce((acc, item) => {
    return (acc += item)
  }, 0).toFixed(2);

  return (
    <>
      <h4>Your Balance</h4>
      <h1>${total}</h1>
    </>
  )
}

export default Balance