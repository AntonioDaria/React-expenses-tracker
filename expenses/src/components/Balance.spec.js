import React from 'react';
import { render, wait } from '@testing-library/react';
import Balance from './Balance.js'
import { GlobalContext, INITIAL_TRANSACTION_CONTEXT_STATE} from '../context/GlobalState'


let dispatch = jest.fn();
describe('Balance component', () => {
    let container;

    it('should match the snapshot', async () => {
      const renderBalance = () => {
        return render(
            <Balance/>
        );
      };
      ({ container } = renderBalance());
      expect(container).toMatchSnapshot();
    });
    

    it('should display the right balance', async () => {
      ({ container } = renderBalanceWithContext());
      await wait();
      expect(container).toMatchSnapshot();
    });

     
});
  

const renderBalanceWithContext = () => {
  let mockTransactions = INITIAL_TRANSACTION_CONTEXT_STATE.transactions
  return render(
    
    <GlobalContext.Provider
    value={{
      dispatch,
      transactions: mockTransactions
    }}
    >
       <Balance/>
    </GlobalContext.Provider>
  );
};
