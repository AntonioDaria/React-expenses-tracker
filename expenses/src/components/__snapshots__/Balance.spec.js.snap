// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Balance component should display the right balance 1`] = `
<div>
  <h4>
    Your Balance
  </h4>
  <h1>
    $
    420.00
  </h1>
</div>
`;

exports[`Balance component should match the snapshot 1`] = `
<div>
  <h4>
    Your Balance
  </h4>
  <h1>
    $
    420.00
  </h1>
</div>
`;
